rfcs/0001-messages.md

```diff
@@ -0,0 +1,414 @@
+/// The data required to serialize the common header fields for every message.
+///
+/// Note: the `msg_type` is derived from the `payload` enum variant.
+struct Header {
+    version: MsgVersion,
+    sender: ParticipantId,
+    receiver: ParticipantId,
+}
+```
+
+While `Payload` will be defined as:
+
+```rust
+/// The data required to serialize the payload for a message.
+enum Payload {
+    SharePackage(messages::SharePackage),
+    SigningCommitments(messages::SigningCommitments),
+    SigningPackage(messages::SigningPackage),
+    SignatureShare(messages::SignatureShare),
+    AggregateSignature(messages::AggregateSignature),
+}
+```
+
+All the messages and new types will be defined in a new file `src/frost/messages.rs`
+
+## Reference-level explanation
+
+Here we explore in detail the header types and all the message payloads. 
+
+### Header
+
+Fields of the header define new types. Proposed implementation for them is as follows:
+
+```rust
+/// The numeric values used to identify each `Payload` variant during serialization.
+#[repr(u8)]
+#[non_exhaustive]
+enum MsgType {
+    SharePackage,
+    SigningCommitments,
+    SigningPackage,
+    SignatureShare,
+    AggregateSignature,
+}
+
+/// The numeric values used to identify the protocol version during serialization.
+struct MsgVersion(u8);
+
+const BASIC_FROST_SERIALIZATION: MsgVersion = MsgVersion(0);
+
+/// The numeric values used to identify each participant during serialization.
+///
+/// In the `frost` module, participant ID `0` should be invalid.
+/// But in serialization, we want participants to be indexed from `0..n`,
+/// where `n` is the number of participants.
+/// This helps us look up their shares and commitments in serialized arrays.
+/// So in serialization, we assign the dealer and aggregator the highest IDs,
+/// and mark those IDs as invalid for signers.
+///
+/// "When performing Shamir secret sharing, a polynomial `f(x)` is used to generate
+/// each party’s share of the secret. The actual secret is `f(0)` and the party with
+/// ID `i` will be given a share with value `f(i)`.
+/// Since a DKG may be implemented in the future, we recommend that the ID `0` be declared invalid."
+/// https://raw.githubusercontent.com/ZcashFoundation/redjubjub/main/zcash-frost-audit-report-20210323.pdf#d
+enum ParticipantId {
+    /// A serialized participant ID for a signer.
+    ///
+    /// Must be less than or equal to `MAX_SIGNER_PARTICIPANT_ID`.
+    Signer(u8),
+    /// The fixed participant ID for the dealer.
+    Dealer,
+    /// The fixed participant ID for the aggregator.
+    Aggregator,
+}
+
+/// The fixed participant ID for the dealer.
+const DEALER_PARTICIPANT_ID: u8 = u8::MAX - 1;
+
+/// The fixed participant ID for the aggregator.
+const AGGREGATOR_PARTICIPANT_ID: u8 = u8::MAX;
+
+/// The maximum `ParticipantId::Signer` in this serialization format.
+///
+/// We reserve two participant IDs for the dealer and aggregator.
+const MAX_SIGNER_PARTICIPANT_ID: u8 = u8::MAX - 2;
```

Comment:
This will be updated after https://github.com/ZcashFoundation/redjubjub/issues/77 is complete, doesn't block this PR.

Reply:
Is this how this will be when we get `u68`?

```
const DEALER_PARTICIPANT_ID: u64 = u64::MAX - 1;
...
```

Reply:
I am askining it because in this case the dealer id will be `18446744073709551614` which is a bit weird.

Reply:
Will it be better if we reserve `u64::MIN` as the dealer and `u64::MAX + 1` as the aggregator ?

Reply:
Yeah the same just `u64::MAX` instead of `u8::MAX`

Reply:
I made changes to FROST at https://github.com/ZcashFoundation/redjubjub/pull/105 and here in the spec at https://github.com/ZcashFoundation/redjubjub/pull/85/commits/9b057f28789f108ae9eb8a984ac25b2dedc8088b